internal/databroker/config_source.go

```diff
@@ -74,108 +88,131 @@ func (src *ConfigSource) rebuild(ctx context.Context, firstTime firstTime) {
 	_, span := trace.StartSpan(ctx, "databroker.config_source.rebuild")
 	defer span.End()
 
-	log.Info(ctx).Msg("databroker: rebuilding configuration")
-
+	now := time.Now()
 	src.mu.Lock()
 	defer src.mu.Unlock()
+	log.Info(ctx).Str("lock-wait", time.Since(now).String()).Msg("databroker: rebuilding configuration")
 
 	cfg := src.underlyingConfig.Clone()
 
 	// start the updater
 	src.runUpdater(cfg)
 
-	seen := map[uint64]string{}
-	for _, policy := range cfg.Options.GetAllPolicies() {
-		id, err := policy.RouteID()
+	now = time.Now()
+	err := src.buildNewConfigLocked(ctx, cfg)
+	if err != nil {
+		log.Error(ctx).Err(err).Msg("databroker: failed to build new config")
+		return
+	}
+	log.Info(ctx).Str("elapsed", time.Since(now).String()).Msg("databroker: built new config")
+
+	src.computedConfig = cfg
+	if !firstTime {
+		src.Trigger(ctx, cfg)
+	}
+
+	metrics.SetConfigInfo(ctx, cfg.Options.Services, "databroker", cfg.Checksum(), true)
+}
+
+func (src *ConfigSource) buildNewConfigLocked(ctx context.Context, cfg *config.Config) error {
+	eg, ctx := errgroup.WithContext(ctx)
+	eg.SetLimit(runtime.NumCPU()/2 + 1)
+	eg.Go(func() error {
+		src.applySettingsLocked(ctx, cfg)
+		err := cfg.Options.Validate()
 		if err != nil {
-			log.Warn(ctx).Err(err).
-				Str("policy", policy.String()).
-				Msg("databroker: invalid policy config, ignoring")
-			return
+			return fmt.Errorf("validating settings: %w", err)
 		}
-		seen[id] = ""
+		return nil
+	})
+	policies := slices.NewSafeSlice[*config.Policy]()
+	for _, cfgpb := range src.dbConfigs {
+		for _, routepb := range cfgpb.GetRoutes() {
+			routepb := routepb
+			eg.Go(func() error {
+				policy, err := src.buildPolicyFromProto(routepb)
+				if err != nil {
+					log.Ctx(ctx).Err(err).Msg("databroker: error building policy from protobuf")
+					return nil
+				}
+				policies.Append(policy)
+				return nil
+			})
+		}
+	}
+	err := eg.Wait()
+	if err != nil {
+		return err
 	}
 
-	var additionalPolicies []config.Policy
+	src.addPolicies(ctx, cfg, policies.Get())
+	return nil
+}
 
+func (src *ConfigSource) applySettingsLocked(ctx context.Context, cfg *config.Config) {
 	ids := maps.Keys(src.dbConfigs)
 	sort.Strings(ids)
 
-	certsIndex := cryptutil.NewCertificatesIndex()
-	for _, cert := range cfg.Options.GetX509Certificates() {
-		certsIndex.Add(cert)
+	var certsIndex *cryptutil.CertificatesIndex
+	if src.enableValidation {
+		certsIndex = cryptutil.NewCertificatesIndex()
+		for _, cert := range cfg.Options.GetX509Certificates() {
+			certsIndex.Add(cert)
+		}
 	}
 
-	// add all the config policies to the list
-	for _, id := range ids {
-		cfgpb := src.dbConfigs[id]
-
+	for i := 0; i < len(ids) && ctx.Err() == nil; i++ {
+		cfgpb := src.dbConfigs[ids[i]]
 		cfg.Options.ApplySettings(ctx, certsIndex, cfgpb.Settings)
-		var errCount uint64
+	}
+}
 
-		err := cfg.Options.Validate()
```

Comment:
I believe `.Validate` has side effects and is necessary for some config options to be usable. I'm not sure what those are though.

Reply:
Oh I see we call it above.